addons/test_mail/tests/test_mail_mail.py

```diff
@@ -65,6 +80,380 @@ def test_mail_mail_return_path(self):
             mail.send()
         self.assertEqual(self._mails[0]['headers']['Return-Path'], '%s@%s' % (self.alias_bounce, self.alias_domain))
 
+    def test_mail_mail_send_exceptions(self):
+        """ Test various use case with exceptions and errors and see how they are
+        managed and stored at mail and notification level. """
+        message = self.test_record.message_post(body='<p>Message', subject='Subject')
+        self.assertFalse(message.notification_ids)
+        mail = self.env['mail.mail'].create([{
+            'body': '<p>Body<:p>',
+            'email_from': False,
+            'email_to': 'test@example.com',
+            'is_notification': True,
+            'subject': 'Subject',
+        }])
+        notification = self.env['mail.notification'].create({
+            'is_read': False,
+            'mail_mail_id': mail.id,
+            'mail_message_id': message.id,
+            'notification_type': 'email',
+            'res_partner_id': self.partner_employee.id,  # not really used for matching except multi-recipients
+        })
+
+        def _reset_data():
```

Comment:
isn't it better to split this test function ? (so you do not need `_reset_data`, and you have one test per exception)